packages/block/block.ts

```diff
@@ -0,0 +1,244 @@
+      for (let k = 0, l = current.edits.length; k < l; ++k) {
+        const edit = current.edits[k]!;
+        if (edit.type === 'block') {
+          edit.block.patch(block.edits?.[i]![k].block);
+          continue;
+        }
+        if (!('hole' in edit) || !edit.hole) continue;
+        const oldValue = props[edit.hole];
+        const newValue = block.props[edit.hole];
+
+        if (Object.is(newValue, oldValue)) continue;
+
+        if (edit.type === 'event') {
+          edit.patch?.(newValue);
+          continue;
+        }
+        if (!el) el = getCurrentElement(current, root, this.cache, i);
+        if (edit.type === 'child') {
+          if (oldValue instanceof AbstractBlock) {
+            // Remember! If we find a block inside a child, we need to locate
+            // the cooresponding block in the new props and patch it.
+            const firstEdit = block.edits?.[i]?.edits[k] as EditChild;
+            const thisSubBlock = block.props[firstEdit.hole];
+            oldValue.patch(thisSubBlock);
+            continue;
+          }
+          setText(el, String(newValue), edit.index);
+        } else if (edit.type === 'attribute') {
+          setAttribute(el, edit.name, newValue);
+        } else if (edit.type === 'style') {
+          setStyleAttribute(el, edit.name, newValue);
+        } else {
+          setSvgAttribute(el, edit.name, newValue);
+        }
+      }
+    }
+
+    return root;
+  }
+  move(block: AbstractBlock | null = null, refNode: Node | null = null) {
+    insertBefore$.call(this.parent, this.el!, block ? block.el! : refNode);
+  }
+  remove() {
+    removeElement$.call(this.el);
+  }
+  shouldUpdate(_oldProps: Props, _newProps: Props): boolean {
+    return true;
+  }
+  toString() {
+    return String(this.el?.outerHTML);
+  }
+  get parent(): HTMLElement | null | undefined {
+    if (!this._parent) this._parent = this.el?.parentElement;
+    return this._parent;
+  }
+}
+
+const getCurrentElement = (
+  current: Edit,
+  root: HTMLElement,
+  cache?: Map<number, HTMLElement>,
+  slot?: number, // edit index
+): HTMLElement => {
+  const pathLength = current.path.length;
+  if (!pathLength) return root;
+  if (cache && slot !== undefined && mapHas$.call(cache, slot)) {
+    return mapGet$.call(cache, slot)!;
+  }
+  // path is an array of indices to traverse the DOM tree
+  // For example, [0, 1, 2] becomes:
+  // root.firstChild.firstChild.nextSibling.firstChild.nextSibling.nextSibling
+  // We use path because we don't have the actual DOM nodes until mount()
+  for (let i = 0; i < pathLength; ++i) {
+    const siblings = current.path[i]!;
+    // https://www.measurethat.net/Benchmarks/Show/15652/0/childnodes-vs-children-vs-firstchildnextsibling-vs-firs
+    root = firstChild$.call(root);
+    if (!siblings) continue;
+    for (let j = 0; j < siblings; ++j) {
+      root = nextSibling$.call(root) as HTMLElement;
+    }
+  }
+  if (cache && slot !== undefined) mapSet$.call(cache, slot, root);
```

Comment:
```suggestion
  const useCache = cache && slot !== undefined;
  if (useCache && mapHas$.call(cache, slot)) {
    return mapGet$.call(cache, slot)!;
  }
  // path is an array of indices to traverse the DOM tree
  // For example, [0, 1, 2] becomes:
  // root.firstChild.firstChild.nextSibling.firstChild.nextSibling.nextSibling
  // We use path because we don't have the actual DOM nodes until mount()
  for (let i = 0; i < pathLength; ++i) {
    const siblings = current.path[i]!;
    // https://www.measurethat.net/Benchmarks/Show/15652/0/childnodes-vs-children-vs-firstchildnextsibling-vs-firs
    root = firstChild$.call(root);
    if (!siblings) continue;
    for (let j = 0; j < siblings; ++j) {
      root = nextSibling$.call(root) as HTMLElement;
    }
  }
  if (useCache) mapSet$.call(cache, slot, root);
```

Reply:
pushed in latest commit!